Revise design engine page (#1317)
ezsystems/developer-documentation: docs/guide/design_engine.md

# Design engine

You can provide design themes for your application, with an automatic fallback system, by using the design engine.

A *design* is a collection of themes. The order of themes within a design is important, because it defines the fallback order.
One design can be used per SiteAccess.

A *theme* is a collection (typically, a folder) of templates and assets. 
For example, templates located under `templates/themes/my_theme` are part of the `my_theme` theme.

When you call a given template or asset, the system looks for it in the first configured theme.
If it cannot be found, the system falls back to other configured themes for the current SiteAccess.

!!! caution

>     After creating a new folder either in a project's or your bundle's `templates/themes` directory,

Reviewer: ```suggestion
    After you create a new folder either in the project's or your bundle's `templates/themes` directory,
```

>     you need to clear Symfony cache (`php bin/console cache:clear`), even if you are working in the `dev` environment.

Reviewer: ```suggestion
    clear the Symfony's cache (`php bin/console cache:clear`), even if you are working in the `dev` environment.
```

### Default designs

> By default the following designs are included:

Reviewer: ```suggestion
By default, the following designs are included:
```

> - `admin` covers templates for the Back Office. It contains the `admin` theme.

Reviewer: ```suggestion
- `admin` - covers templates for the Back Office, contains the `admin` theme
```

> - `standard` covers the default content rendering templates. It contains the `standard` theme.

Reviewer: ```suggestion
- `standard` - covers the default content rendering templates, contains the `standard` theme
```

> - `base_design` covers the shop templates. It contains `base_theme`.

Reviewer: ```suggestion
- `base_design` - covers the shop templates, contains `base_theme`
```

- `content_fields.html.twig`
- `fielddefinition_settings.html.twig`

## Adding a design

To define and use a design, you need to:

1. Declare it, with a name and a collection of themes to use.
1. Use it for your SiteAccess.

Define your design under the `ezdesign.design_list` key:

``` yaml
ezdesign:
    design_list:
        my_design: [theme1, theme2]
```

> The design engine tries `theme1` first. If it cannot find a template in `theme1`, it tries `theme2`.

Reviewer: ```suggestion
The design engine tries `theme1` first. 
If it cannot find a template in `theme1`, it tries `theme2`.
```

$currentDesign = $this->getConfigResolver->getParameter('design');
```

## Using designs

By convention, a theme directory is located under `templates/themes/<theme_name>`.

> To indicate a template from a theme, use the `@ezdesign` special Twig namespace.

Reviewer: ```suggestion
To indicate a template from a theme, use the `@ezdesign` Twig namespace.
```

By convention, an asset theme directory is located in `assets/themes/<theme_name>`.

It is also possible to use `assets` as a global override directory.
If a called asset is present directly under this directory, it is always considered first.

!!! caution

    You must have *installed* your assets with `assets:install` command, so that your public resources are
    *installed* into the `assets/` directory.

### Fallback order

The default fallback order for templates is:

- Application theme directory: `templates/themes/<theme_name>`
- Bundle theme directory: `src/<bundle_directory>/Resources/views/themes/<theme_name>`

The default fallback order for assets is:

- Application assets directory: `assets`
- Application theme directory: `assets/themes/<theme_name>`
- Bundle theme directory: `public/bundles/<bundle_directory>/themes/<theme_name>`

Calling `asset("js/scripts.js", "ezdesign")` can for example be resolved to `public/bundles/app/themes/my_theme/js/scripts.js`.

Bundles fall back in the instantiation order from `bundles.php`.

#### Additional theme paths

You can also add any Twig template folder to the theme configuration.

> You can use it if you want to define templates from third-party bundles as part of one of your theme,

Reviewer: ```suggestion
You can use it if you want to define templates from third-party bundles as part of one of your themes,
```

> or when upgrading your application in order to use the design engine,

Reviewer: ```suggestion
or when upgrading your application to use the design engine,
```

> if your existing templates are not yet following the convention.

Reviewer: if -> when ?

> if your existing templates are not yet following the convention.

Author: ```suggestion
and if your existing templates are not yet following the convention.
```
It's actually upgrading+existing template not following convention, so maybe this way to clarify it's not A or B or C.

Do it by setting the `ezdesign.templates_theme_paths` parameter:

```yaml
ezdesign:
    design_list:
        my_design: [my_theme]
    templates_theme_paths:
        my_theme:
            - '%kernel.project_dir%/vendor/<vendor_name>/<bundle_name>/Resources/views'
```

> Directories following the convention always have precedence over the ones defined in `templates_theme_paths`.

Reviewer: ```suggestion
Directories that follow the convention have precedence over the ones defined in `templates_theme_paths`.
```

This ensures that it is always possible to override a template from the application.

You can also add a global override folder, by listing paths without assigning them to a theme:

```yaml
ezdesign:
    templates_override_paths:
        - '%kernel.project_dir%/src/an_override_directory'
```

### Performance and asset resolution

When using themes, paths for assets are resolved at runtime.
This is due to how the Symfony Asset component is integrated with Twig.
This can cause significant performance impact because of I/O calls when looping over all potential theme directories,
especially when using a lot of different designs and themes.

To work around this issue, assets resolution can be provisioned at compilation time.

> Provisioning is the default behavior in non-debug mode (e.g. `prod` environment).

Reviewer: ```suggestion
Provisioning is the default behavior in non-debug mode (for example, in `prod` environments).
```

> In debug mode (e.g. `dev` environment), assets are being resolved at runtime.

Reviewer: ```suggestion
In debug mode (for example, in `dev` environments), assets are being resolved at runtime.
```
